odl/deform/mass_preserving.py

```diff
@@ -0,0 +1,548 @@
+    >>> deform_field = deform_field_space.element([[0.1, 0.3, 0.5, 0.6, 0.9]])
+    >>> _general_deform(template, deform_field)
+    array([ 0. ,  0. ,  1. ,  0.5,  0. ])
+    """
+    image_pts = template.space.points()
+    for i, vi in enumerate(deformation):
+        image_pts[:, i] = vi.ntuple.asarray()
+    return template.interpolation(image_pts.T, out=out, bounds_check=False)
+
+
+def _mass_presv_deform(template, mass_presv_dfield, out=None):
+    """Deform a template with a mass-preserving deformation field.
+
+    The function maps a given template ``I``, a mass-preserving deformation
+    field, i.e., a given deformation field ``phi`` and its Jacobian
+    determinant to the new function ``x --> |Dphi(x)|I(phi(x))``.
+
+    Parameters
+    ----------
+    template : `DiscreteLpVector`
+        Template to be deformed by a deformation field.
+    mass_presv_dfield : element of power space of power space and `DiscreteLp`
+        The mass-preserving deformation field used to deform the template.
+        It contains the deformation, i.e., an element of power space
+        of `DiscreteLp`, and the Jacobian determinant of the deformation,
+        i.e., a `DiscreteLpVector`.
+    out : `numpy.ndarray`, optional
+        Array to which the function values of the deformed template
+        are written. It must have the same shape as ``template`` and
+        a data type compatible with ``template.dtype``.
+
+    Returns
+    -------
+    deformed_template : `numpy.ndarray`
+        Function values of the deformed template. If ``out`` was given,
+        the returned object is a reference to it.
+
+    Examples
+    --------
+    Create a simple 1D template ``I`` over ``[0, 1]``, a deffeomorphic
+    deformation field and its Jacobian determinant.
+    If the deffeomorphism ``phi`` is ``sin(x*pi/2)``, its Jacobian determinant
+    should be ``cos(x*pi/2)*pi/2`` and the output value is
+    ``cos(x*pi/2)*pi/2*I(sin(x*pi/2))``. A mass-preserving deformation
+    is obtained, namely, the integration of ``I(x)`` over ``[0, 1]``
+    equals to the integration of ``|Dphi(x)|I(phi(x))`` over ``[0, 1]``.
+
+
+    >>> import odl
+    >>> import numpy as np
+    >>> space = odl.uniform_discr(0, 1, 5, interp='linear')
+    >>> vspace = space.vector_field_space
+    >>> pspace = odl.ProductSpace(vspace, space)
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> deform_field = vspace.element([np.sin(space.points() * np.pi / 2.0)])
+    >>> det = space.element(np.cos(space.points() * np.pi / 2.0) * np.pi / 2.0)
+    >>> mass_presv_field = pspace.element()
+    >>> mass_presv_field[0] = deform_field
+    >>> mass_presv_field[1] = det
+    >>> print(_mass_presv_deform(template, mass_presv_field))
+    [0.0, 1.07761764468, 0.0, 0.0, 0.0]
+    """
+    general_deform_template = _general_deform(template, mass_presv_dfield[0])
+    return mass_presv_dfield[1] * general_deform_template
+
+
+class GeneralDeformFixedTempl(Operator):
```

Comment:
I'm sceptical to this design. The name suggests that this operator is supposed to perform "deformations in general", for "any deformation model that we choose". 

However, we never **explicitly represent** deformations, or, to put it more clearly, we don't have a type (class) that represents the **elements of some deformation (semi-) group**. 
In contrast, we directly represent **the group action** by a deformation operator and use a **vector field** instead to *implicitly* represent a deformation. 
As we learned from discussions with Klas, basically all deformation models are in some way based on vector fields, so that choice should not limit us in any way. Perhaps we need to revise the decision later, but now I don't see a reason for it.

So, in conclusion, I don't see a good reason to have a "general deformation operator", since our operators encompass both the steps "compute deformation from vector field" and "apply deformation", and the combination of those is *unique for each deformation model*.